File: java/ql/integration-tests/kotlin/all-platforms/nested_generic_types/libsrc/extlib/OuterGeneric.java
package extlib;

public class OuterGeneric<T> {

  public class InnerNotGeneric {

    public T identity(T t) { return t; }

  }

  public InnerNotGeneric getInnerNotGeneric() { return null; }

  public class InnerGeneric<S> {

    public <R> InnerGeneric(R r) { }

    public <R> InnerGeneric(R r, S s) { }

    public S returnsecond(T t, S s) { return s; }

    public <R> S returnsecond(T t, S s, R r) { return s; }

  }

  public static class InnerStaticGeneric<S> {

    public <R> InnerStaticGeneric(R r, S s) { }

    public S identity(S s) { return s; }

  }

}

File: java/ql/test/kotlin/library-tests/generic-inner-classes/OuterGeneric.kt
package testuser

public class OuterGeneric<T> {

  public inner class InnerNotGeneric {

    fun identity(t: T): T { return t }

  }

  public inner class InnerGeneric<S> {

    constructor() { }

    constructor(s: S) { }

    fun returnsecond(t: T, s: S): S { return s; }

  }

}

